src/pest_parser/hash.pest

```diff
@@ -0,0 +1,429 @@
+
+/// Exhaustive patterns are those that are guaranteed to have a finite
+/// number of variants (level 1)
+exhaustive_pattern = {
+    struct_pattern    |
+    ident             | // binding pattern
+    namespace_pattern |
+    ("(" ~ (name ~ (comma ~ name)*)? ~ ")") // tuple pattern
+}
+
+
+braced_pattern = { "(" ~ pattern ~ ")" }
+
+/// Single pattern combinator is used as a combinator for match
+/// statements where multiple patterns can be specified for a 
+/// switch case. 
+single_pattern = { 
+    braced_pattern     |
+    enum_pattern       |
+    exhaustive_pattern |
+    literal
+}
+
+/// The pattern grammar
+pattern = {single_pattern ~ ("if" ~ unary_expr)? ~ ("|" ~ pattern)?}
+
+/// Conditional if expression grammar
+if_statement = {"if" ~ expr ~ block ~ ("else" ~ (if_statement | block))?}
+
+/// Conditional match expression grammar
+match_case      = {pattern ~ "=>" ~ expr}
+match_statement = {"match" ~ expr ~ "{" ~ (match_case ~ semi)* ~ "}"}
+
+/// Body block grammar, body blocks can be used within any kind
+/// of expression to denote a body of a particular expression.
+/// The expression at the end is used as a implicit 'return' statement
+/// from the given body block.
+body_block = { "{" ~  statement* ~ expr? ~ "}" }
+
+/// Block
+block = { 
+    if_statement    |
+    match_statement |
+    for_statement   |
+    while_statement | 
+    loop_statement  |
+    body_block
+}
+
+/// Grammar representation for a literals
+literal = {
+    string        |
+    int           |
+    float         |
+    char          |
+    list_literal  |
+    tuple_literal |
+    map_literal   |
+    set_literal   |
+    function_literal
+}
+
+
+/// Grammar representation of some logical operation involving a 
+/// unary or a binary expression
+binary_expr = {expr_inner ~ (binary_operator ~ expr_inner)+ }
+unary_expr = {unary_operator ~ expr}
+
+/// Grammar represenation for an expression which is a single part of a statement.
+/// The 'inner' part is a workaround for not being able to use LR grammars.
+expr_inner = {
+    literal         |
+    function_call   |
+    unary_expr      |
+    intrinsic       |
+    access_name     | // variable
+    property_access |
+    block           |
+    import
+}
+
+/// Grammar for an expression, can either be a binary expression which refers
+/// to two expressions joined by a binary operator, or a single expression
+/// that can be some other variant of an expression.
+expr = _{binary_expr | expr_inner}
+
+
+return_statement   = { "return" ~ expr? }
+break_statement    = { "break" }
+continue_statement = { "continue" }
+
+/// Statement
+statement = {
+    (decl              |
+    block              |
+    return_statement   |
+    break_statement    |
+    continue_statement |
+    struct_def         |
+    enum_def           |
+    trait_def          |
+    assign             |
+    expr ) ~ ";"
+}
+
+/// Module grammar
+module = _{ SOI ~ "\n"* ~ (statement ~ "\n"+) * ~ statement? ~ EOI }
```

Comment:
Why separated by newlines?

Reply:
Statements could be separated some number of newlines, I think it would be better to change this to whitespace